sematic/resolvers/state_machine_resolver.py

```diff
@@ -183,6 +185,72 @@ def _run_inline(self, future: AbstractFuture) -> None:
     def _wait_for_scheduled_runs(self) -> None:
         pass
 
+    def _fail_future_with_timeout(
+        self,
+        executing_future: AbstractFuture,
+        timeout_restricting_future: typing.Optional[AbstractFuture] = None,
+    ):
+        if timeout_restricting_future is None:
+            timeout_restricting_future = executing_future
+
+        if timeout_restricting_future.id == executing_future.id:
+            message = (
+                f"The Sematic function {executing_future.calculator.__name__} did not "
+                f"complete in time to satisfy its "
+                f"{executing_future.props.timeout_minutes} minute timeout."
+            )
+        else:
+            message = (
+                f"The Sematic function {executing_future.calculator.__name__} did not "
+                f"complete in time to satisfy the "
+                f"{timeout_restricting_future.props.timeout_minutes} minute timeout "
+                f"specified by its parent "
+                f"{timeout_restricting_future.calculator.__name__}."
+            )
+        self._handle_future_failure(
+            executing_future,
+            format_exception_for_run(TimeoutError(message)),
+        )
+
+    def _wait_for_scheduled_runs_with_timeout(self) -> None:
+        seconds_until_next_timeout, future = self._get_seconds_to_next_timeout()
+        if seconds_until_next_timeout is not None and future is not None:
+            logger.info(
+                "Next timeout is in %s seconds, for %s",
+                seconds_until_next_timeout,
+                future.id,
+            )
+            if seconds_until_next_timeout <= 0 and not future.state.is_terminal():
+                # a future had already expired before we even started waiting.
+                self._fail_future_with_timeout(future)
+                return
+
+        try:
+            with timeout(seconds_until_next_timeout):
+                self._wait_for_scheduled_runs()
+        except TimeoutError:
+            if future is not None and not future.state.is_terminal():
+                self._fail_future_with_timeout(future)
+            elif future is not None:
+                logger.warning(
+                    "Received TimeoutError, but %s is already in terminal state %s",
+                    future.id,
+                    future.state,
+                )
+
+    def _get_seconds_to_next_timeout(
```

Comment:
`_get_seconds_to_next_timeout` and `get_minimum_call_chain_timeout_seconds` duplicate the main logic. Please consider replacing them with invocations of a deduplicated function that takes the list of futures as a parameter. `_get_seconds_to_next_timeout` can take `self._futures` and `get_minimum_call_chain_timeout_seconds` can take the result of a new property on `AbstractFuture` that builds its inheritance chain.

Reply:
done